unicode_regex.h

```diff
@@ -0,0 +1,252 @@
+#pragma once
+
+#include "unicode.h"
+#include "unordered_set"
+
+class llm_regex {
+public:
+    std::vector<std::string> gpt2_style(const std::string & str) {
+        std::vector<std::string> results;
+        results.reserve(str.size());
+
+        auto codepoints = unicode_engine.to_codepoints(str);
+
+        for (auto & cp : gpt2_style_implement(codepoints)) {
+            results.push_back(unicode_engine.to_string(cp));
+        }
+
+        return results;
+    }
+
+    std::vector<std::string> falcon_style(const std::string & str) {
+        std::vector<std::string> results;
+        results.reserve(str.size());
+
+        auto codepoints = unicode_engine.to_codepoints(str);
+
+        for (auto & cp_1 : split_punctuation_unicode_ascii(codepoints)) {
+            for (auto & cp_2 : gpt2_style_implement(cp_1)) {
+                for (auto & cp_3 : split_digits_unicode(cp_2)) {
+                    for (auto & cp_4 : split_continuous_digits_ascii(cp_3)) {
+                        results.push_back(unicode_engine.to_string(cp_4));
+                    }
+                }
+            }
+        }
+
+        return results;
+    }
+
+    UNICODE & get_unicode_engine() {
+        return unicode_engine;
+    }
+
+    llm_regex() {
+        unicode_engine.overload_category(REGEX_RANGES::Whitespace, "WHITESPACE");

```

Comment:
Why is this done explicitly instead of being done by default?

Reply:
In developing a universal Unicode engine, I face a notable challenge: the `Unicode 15.0` standard, as defined by the Unicode Consortium, does not specify a definition for whitespace. Furthermore, the interpretation of `\s` whitespace can vary among different regex engines.